src/branch.rs

```diff
@@ -4,255 +4,283 @@
-        }
+        &*self.node
     }
 }
 
-impl<'a, C, S> Level<'a, C, S>
-where
-    C: Clone,
-{
-    /// Returns the offset of the branch level
-    pub fn offset(&self) -> usize {
-        self.offset
-    }
-
-    fn new_owned(node: C) -> Self {
+impl<'a, C, A> Level<'a, C, A> {
+    pub fn new_root(root: &'a C) -> Level<'a, C, A> {
         Level {
             offset: 0,
-            inner: LevelInner::Owned(node, PhantomData),
+            node: LevelNode::Root(root),
         }
     }
 
-    fn new_borrowed(node: &'a C) -> Self {
+    pub fn new_val(ann: AnnRef<'a, C, A>) -> Level<'a, C, A> {
         Level {
             offset: 0,
-            inner: LevelInner::Borrowed(node),
+            node: LevelNode::Val(ann),
         }
     }
 
-    fn offset_mut(&mut self) -> &mut usize {
-        &mut self.offset
+    /// Returns the offset of the branch level
+    pub fn offset(&self) -> usize {
+        self.offset
     }
 
-    fn inner(&self) -> &LevelInner<'a, C, S> {
-        &self.inner
+    fn offset_mut(&mut self) -> &mut usize {
+        &mut self.offset
     }
 }
 
-#[derive(Clone)]
-enum LevelInner<'a, C, S>
-where
-    C: Clone,
-{
-    Borrowed(&'a C),
-    Owned(C, PhantomData<S>),
-}
+#[derive(Debug)]
+pub struct PartialBranch<'a, C, A>(Vec<Level<'a, C, A>>);
 
-pub struct PartialBranch<'a, C, S>(Levels<'a, C, S>)
+impl<'a, C, A> Deref for LevelNode<'a, C, A>
 where
-    C: Clone;
+    C: Compound<A>,
+    A: Annotation<C::Leaf>,
+{
+    type Target = C;
 
-pub struct Levels<'a, C, S>(Vec<Level<'a, C, S>>)
-where
-    C: Clone;
+    fn deref(&self) -> &Self::Target {
+        match self {
+            LevelNode::Root(target) => target,
+            LevelNode::Val(val) => &**val,
+        }
+    }
+}
 
-impl<'a, C, S> Levels<'a, C, S>
+impl<'a, C, A> PartialBranch<'a, C, A>
 where
-    C: Compound<S>,
-    S: Store,
+    C: Compound<A>,
+    A: Annotation<C::Leaf>,
 {
-    pub fn new(node: &'a C) -> Self {
-        let mut levels: Vec<Level<'a, C, S>> = Vec::new();
-        levels.push(Level::new_borrowed(node));
-        Levels(levels)
+    fn new(root: &'a C) -> Self {
+        let mut vec = Vec::new();
+        vec.push(Level::new_root(root));
+        PartialBranch(vec)
     }
 
     pub fn depth(&self) -> usize {
         self.0.len()
     }
 
-    pub fn top(&self) -> &Level<'a, C, S> {
-        self.0.last().expect("always > 0 len")
-    }
-
-    pub fn top_mut(&mut self) -> &mut Level<'a, C, S> {
-        self.0.last_mut().expect("always > 0 len")
-    }
+    fn leaf(&self) -> Option<&'a C::Leaf> {
+        let top = self.top();
+        let ofs = top.offset();
 
-    pub fn pop(&mut self) -> bool {
-        if self.0.len() > 1 {
-            self.0.pop();
-            true
-        } else {
-            false
+        match top.child(ofs) {
+            Child::Leaf(l) => {
+                let l: &C::Leaf = l;
+                let l_extended: &'a C::Leaf =
+                    unsafe { core::mem::transmute(l) };
```

Comment:
I thought we weren't going to use `unsafe` in microkelvin, since it was the original reason why we move from kelvin.

Considering that we don't care (currently) much about the performance, can we avoid the usage of `unsafe` in microkelvin at all? If not, can we add a comment why?

Reply:
good catch